compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/SyntheticAccessorLowering.kt

```diff
@@ -150,38 +150,38 @@ private class SyntheticAccessorLowering(val context: JvmBackendContext) : IrElem
             copyAllParamsToArgs(it, accessor)
         }
 
-    private fun IrSimpleFunction.makeSimpleFunctionAccessor(superQualifierSymbol: IrClassSymbol?): IrSimpleFunction {
+    private fun IrSimpleFunction.makeSimpleFunctionAccessor(expression: IrCall): IrSimpleFunction {
         val source = this
         return buildFun {
             origin = JvmLoweredDeclarationOrigin.SYNTHETIC_ACCESSOR
             name = source.accessorName()
             visibility = Visibilities.PUBLIC
             isSuspend = false // do not generate state-machine for synthetic accessors
         }.also { accessor ->
-            accessor.parent = if ((source.parent as? IrClass)?.isInterface == true) {
+            // Resolve the actual parent for a super call
+            val parent = expression.superQualifierSymbol?.let { expression.dispatchReceiver!!.type.classOrNull!!.owner } ?: parent
```

Comment:
This code seems pretty convoluted. I'm having a hard time figuring out what is happening here. Maybe you can expand the comment to explain what the receiver of makeSimpleFunctionAccessor is in these two cases and show by examples why this parent computation does the right thing?

Reply:
I've simplified the code and added some comments. To be honest, this code *is* fragile and currently relies on having correct IR types. I don't know how to find the correct enclosing class for super calls otherwise.